lib/libzpool/uzfs_io.c

```diff
@@ -172,32 +172,29 @@ uzfs_write_data(zvol_state_t *zv, char *buf, uint64_t offset, uint64_t len,
 /* Reads data from volume 'zv', and fills up memory at buf */
 int
 uzfs_read_data(zvol_state_t *zv, char *buf, uint64_t offset, uint64_t len,
-    void **md, uint64_t *mdlen)
+    metadata_desc_t **md_head)
 {
-	int error = 0;
+	int error = EINVAL;	// in case we aren't able to read a single block
 	uint64_t blocksize = zv->zv_volblocksize;
 	uint64_t bytes = 0;
 	uint64_t volsize = zv->zv_volsize;
 	uint64_t end = len + offset;
 	uint64_t read = 0;
 	objset_t *os = zv->zv_objset;
 	rl_t *rl, *mrl;
-	int ret = 0;
 	uint64_t r_offset;
 	uint64_t metadatasize = zv->zv_volmetadatasize;
-	void *mdata = NULL;
-	uint64_t mread = 0;
-	uint64_t mlen = 0;
 	metaobj_blk_offset_t metablk;
 	uint64_t len_in_first_aligned_block = 0;
+	metadata_desc_t *md_ent = NULL, *new_md;
 
-	if (zv->zv_volmetablocksize == 0)
-		mdlen = NULL;
+	ASSERT3P(zv->zv_volmetadatasize, ==, sizeof (blk_metadata_t));
 
-	if (md != NULL && mdlen != NULL) {
-		mlen = get_metadata_len(zv, offset, len);
-		mdata = kmem_alloc(mlen, KM_SLEEP);
-		mread = 0;
+	/* init metadata in case caller wants to receive that info */
+	if (md_head != NULL) {
+		*md_head = NULL;
+		if (zv->zv_volmetablocksize == 0)
```

Comment:
this should never happen. ASSERT can be added.

Reply:
you are right. We check that zvol has metadata when we open it and otherwise refuse to create zinfo for it. I will change it to assert.